File: runker/src/main/java/com/irun/runker/net/util/MediaFile.java
package com.irun.runker.net.util;

import android.annotation.SuppressLint;
import android.mtp.MtpConstants;

import java.io.File;
import java.util.HashMap;

@SuppressWarnings({"WeakerAccess", "unused"})
public class MediaFile {

    private MediaFile() {
    }

    // Audio file types
    public static final int FILE_TYPE_MP3 = 1;
    public static final int FILE_TYPE_M4A = 2;
    public static final int FILE_TYPE_WAV = 3;
    public static final int FILE_TYPE_AMR = 4;
    public static final int FILE_TYPE_AWB = 5;
    public static final int FILE_TYPE_WMA = 6;
    public static final int FILE_TYPE_OGG = 7;
    public static final int FILE_TYPE_AAC = 8;
    public static final int FILE_TYPE_MKA = 9;
    public static final int FILE_TYPE_FLAC = 10;
    private static final int FIRST_AUDIO_FILE_TYPE = FILE_TYPE_MP3;
    private static final int LAST_AUDIO_FILE_TYPE = FILE_TYPE_FLAC;
    // MIDI file types
    public static final int FILE_TYPE_MID = 11;
    public static final int FILE_TYPE_SMF = 12;
    public static final int FILE_TYPE_IMY = 13;
    private static final int FIRST_MIDI_FILE_TYPE = FILE_TYPE_MID;
    private static final int LAST_MIDI_FILE_TYPE = FILE_TYPE_IMY;

    // Video file types
    public static final int FILE_TYPE_MP4 = 21;
    public static final int FILE_TYPE_M4V = 22;
    public static final int FILE_TYPE_3GPP = 23;
    public static final int FILE_TYPE_3GPP2 = 24;
    public static final int FILE_TYPE_WMV = 25;
    public static final int FILE_TYPE_ASF = 26;
    public static final int FILE_TYPE_MKV = 27;
    public static final int FILE_TYPE_MP2TS = 28;
    public static final int FILE_TYPE_AVI = 29;
    public static final int FILE_TYPE_WEBM = 30;
    private static final int FIRST_VIDEO_FILE_TYPE = FILE_TYPE_MP4;
    private static final int LAST_VIDEO_FILE_TYPE = FILE_TYPE_WEBM;

    // More video file types
    public static final int FILE_TYPE_MP2PS = 200;
    private static final int FIRST_VIDEO_FILE_TYPE2 = FILE_TYPE_MP2PS;
    private static final int LAST_VIDEO_FILE_TYPE2 = FILE_TYPE_MP2PS;
    // Image file types
    public static final int FILE_TYPE_JPEG = 31;
    public static final int FILE_TYPE_GIF = 32;
    public static final int FILE_TYPE_PNG = 33;
    public static final int FILE_TYPE_BMP = 34;
    public static final int FILE_TYPE_WBMP = 35;
    public static final int FILE_TYPE_WEBP = 36;
    private static final int FIRST_IMAGE_FILE_TYPE = FILE_TYPE_JPEG;
    private static final int LAST_IMAGE_FILE_TYPE = FILE_TYPE_WEBP;

    // Playlist file types
    public static final int FILE_TYPE_M3U = 41;
    public static final int FILE_TYPE_PLS = 42;
    public static final int FILE_TYPE_WPL = 43;
    public static final int FILE_TYPE_HTTPLIVE = 44;
    private static final int FIRST_PLAYLIST_FILE_TYPE = FILE_TYPE_M3U;
    private static final int LAST_PLAYLIST_FILE_TYPE = FILE_TYPE_HTTPLIVE;
    // Drm file types
    public static final int FILE_TYPE_FL = 51;
    private static final int FIRST_DRM_FILE_TYPE = FILE_TYPE_FL;
    private static final int LAST_DRM_FILE_TYPE = FILE_TYPE_FL;
    // Other popular file types
    public static final int FILE_TYPE_TEXT = 100;
    public static final int FILE_TYPE_HTML = 101;
    public static final int FILE_TYPE_PDF = 102;
    public static final int FILE_TYPE_XML = 103;
    public static final int FILE_TYPE_MS_WORD = 104;
    public static final int FILE_TYPE_MS_EXCEL = 105;
    public static final int FILE_TYPE_MS_POWERPOINT = 106;
    public static final int FILE_TYPE_ZIP = 107;

    public static class MediaFileType {
        public final int fileType;
        public final String mimeType;

        MediaFileType(int fileType, String mimeType) {
            this.fileType = fileType;
            this.mimeType = mimeType;
        }
    }

    private static final HashMap<String, MediaFileType> sFileTypeMap = new HashMap<>();
    private static final HashMap<String, Integer> sMimeTypeMap = new HashMap<>();
    // maps file extension to MTP format code
    private static final HashMap<String, Integer> sFileTypeToFormatMap = new HashMap<>();
    // maps mime type to MTP format code
    private static final HashMap<String, Integer> sMimeTypeToFormatMap = new HashMap<>();
    // maps MTP format code to mime type
    @SuppressLint("UseSparseArrays")
    private static final HashMap<Integer, String> sFormatToMimeTypeMap = new HashMap<>();

    static void addFileType(String extension, int fileType, String mimeType) {
        sFileTypeMap.put(extension, new MediaFileType(fileType, mimeType));
        sMimeTypeMap.put(mimeType, fileType);
    }

    static void addFileType(String extension, int fileType, String mimeType, int mtpFormatCode) {
        addFileType(extension, fileType, mimeType);
        sFileTypeToFormatMap.put(extension, mtpFormatCode);
        sMimeTypeToFormatMap.put(mimeType, mtpFormatCode);
        sFormatToMimeTypeMap.put(mtpFormatCode, mimeType);
    }

    static {
        addFileType("MP3", FILE_TYPE_MP3, "audio/mpeg", MtpConstants.FORMAT_MP3);
        addFileType("MPGA", FILE_TYPE_MP3, "audio/mpeg", MtpConstants.FORMAT_MP3);
        addFileType("M4A", FILE_TYPE_M4A, "audio/mp4", MtpConstants.FORMAT_MPEG);
        addFileType("WAV", FILE_TYPE_WAV, "audio/x-wav", MtpConstants.FORMAT_WAV);
        addFileType("AMR", FILE_TYPE_AMR, "audio/amr");
        addFileType("AWB", FILE_TYPE_AWB, "audio/amr-wb");

        addFileType("OGG", FILE_TYPE_OGG, "audio/ogg", MtpConstants.FORMAT_OGG);
        addFileType("OGG", FILE_TYPE_OGG, "application/ogg", MtpConstants.FORMAT_OGG);
        addFileType("OGA", FILE_TYPE_OGG, "application/ogg", MtpConstants.FORMAT_OGG);
        addFileType("AAC", FILE_TYPE_AAC, "audio/aac", MtpConstants.FORMAT_AAC);
        addFileType("AAC", FILE_TYPE_AAC, "audio/aac-adts", MtpConstants.FORMAT_AAC);
        addFileType("MKA", FILE_TYPE_MKA, "audio/x-matroska");

        addFileType("MID", FILE_TYPE_MID, "audio/midi");
        addFileType("MIDI", FILE_TYPE_MID, "audio/midi");
        addFileType("XMF", FILE_TYPE_MID, "audio/midi");
        addFileType("RTTTL", FILE_TYPE_MID, "audio/midi");
        addFileType("SMF", FILE_TYPE_SMF, "audio/sp-midi");
        addFileType("IMY", FILE_TYPE_IMY, "audio/imelody");
        addFileType("RTX", FILE_TYPE_MID, "audio/midi");
        addFileType("OTA", FILE_TYPE_MID, "audio/midi");
        addFileType("MXMF", FILE_TYPE_MID, "audio/midi");

        addFileType("MPEG", FILE_TYPE_MP4, "video/mpeg", MtpConstants.FORMAT_MPEG);
        addFileType("MPG", FILE_TYPE_MP4, "video/mpeg", MtpConstants.FORMAT_MPEG);
        addFileType("MP4", FILE_TYPE_MP4, "video/mp4", MtpConstants.FORMAT_MPEG);
        addFileType("M4V", FILE_TYPE_M4V, "video/mp4", MtpConstants.FORMAT_MPEG);
        addFileType("3GP", FILE_TYPE_3GPP, "video/3gpp", MtpConstants.FORMAT_3GP_CONTAINER);
        addFileType("3GPP", FILE_TYPE_3GPP, "video/3gpp", MtpConstants.FORMAT_3GP_CONTAINER);
        addFileType("3G2", FILE_TYPE_3GPP2, "video/3gpp2", MtpConstants.FORMAT_3GP_CONTAINER);
        addFileType("3GPP2", FILE_TYPE_3GPP2, "video/3gpp2", MtpConstants.FORMAT_3GP_CONTAINER);
        addFileType("MKV", FILE_TYPE_MKV, "video/x-matroska");
        addFileType("WEBM", FILE_TYPE_WEBM, "video/webm");
        addFileType("TS", FILE_TYPE_MP2TS, "video/mp2ts");
        addFileType("AVI", FILE_TYPE_AVI, "video/avi");

        addFileType("JPG", FILE_TYPE_JPEG, "image/jpeg", MtpConstants.FORMAT_EXIF_JPEG);
        addFileType("JPEG", FILE_TYPE_JPEG, "image/jpeg", MtpConstants.FORMAT_EXIF_JPEG);
        addFileType("GIF", FILE_TYPE_GIF, "image/gif", MtpConstants.FORMAT_GIF);
        addFileType("PNG", FILE_TYPE_PNG, "image/png", MtpConstants.FORMAT_PNG);
        addFileType("BMP", FILE_TYPE_BMP, "image/x-ms-bmp", MtpConstants.FORMAT_BMP);
        addFileType("WBMP", FILE_TYPE_WBMP, "image/vnd.wap.wbmp");
        addFileType("WEBP", FILE_TYPE_WEBP, "image/webp");

        addFileType("M3U", FILE_TYPE_M3U, "audio/x-mpegurl", MtpConstants.FORMAT_M3U_PLAYLIST);
        addFileType("M3U", FILE_TYPE_M3U, "application/x-mpegurl", MtpConstants.FORMAT_M3U_PLAYLIST);
        addFileType("PLS", FILE_TYPE_PLS, "audio/x-scpls", MtpConstants.FORMAT_PLS_PLAYLIST);
        addFileType("WPL", FILE_TYPE_WPL, "application/vnd.ms-wpl", MtpConstants.FORMAT_WPL_PLAYLIST);
        addFileType("M3U8", FILE_TYPE_HTTPLIVE, "application/vnd.apple.mpegurl");
        addFileType("M3U8", FILE_TYPE_HTTPLIVE, "audio/mpegurl");
        addFileType("M3U8", FILE_TYPE_HTTPLIVE, "audio/x-mpegurl");
        addFileType("FL", FILE_TYPE_FL, "application/x-android-drm-fl");
        addFileType("TXT", FILE_TYPE_TEXT, "text/plain", MtpConstants.FORMAT_TEXT);
        addFileType("HTM", FILE_TYPE_HTML, "text/html", MtpConstants.FORMAT_HTML);
        addFileType("HTML", FILE_TYPE_HTML, "text/html", MtpConstants.FORMAT_HTML);
        addFileType("PDF", FILE_TYPE_PDF, "application/pdf");
        addFileType("DOC", FILE_TYPE_MS_WORD, "application/msword", MtpConstants.FORMAT_MS_WORD_DOCUMENT);
        addFileType("XLS", FILE_TYPE_MS_EXCEL, "application/vnd.ms-excel", MtpConstants.FORMAT_MS_EXCEL_SPREADSHEET);
        addFileType("PPT", FILE_TYPE_MS_POWERPOINT, "application/mspowerpoint", MtpConstants.FORMAT_MS_POWERPOINT_PRESENTATION);
        addFileType("FLAC", FILE_TYPE_FLAC, "audio/flac", MtpConstants.FORMAT_FLAC);
        addFileType("ZIP", FILE_TYPE_ZIP, "application/zip");
        addFileType("MPG", FILE_TYPE_MP2PS, "video/mp2p");
        addFileType("MPEG", FILE_TYPE_MP2PS, "video/mp2p");
    }

    public static boolean isAudioFileType(int fileType) {
        return ((fileType >= FIRST_AUDIO_FILE_TYPE && fileType <= LAST_AUDIO_FILE_TYPE) ||
                (fileType >= FIRST_MIDI_FILE_TYPE && fileType <= LAST_MIDI_FILE_TYPE));
    }

    public static boolean isVideoFileType(int fileType) {
        return (fileType >= FIRST_VIDEO_FILE_TYPE && fileType <= LAST_VIDEO_FILE_TYPE)
                || (fileType >= FIRST_VIDEO_FILE_TYPE2 && fileType <= LAST_VIDEO_FILE_TYPE2);
    }

    public static boolean isImageFileType(int fileType) {
        return (fileType >= FIRST_IMAGE_FILE_TYPE && fileType <= LAST_IMAGE_FILE_TYPE);
    }

    public static boolean isPlayListFileType(int fileType) {
        return (fileType >= FIRST_PLAYLIST_FILE_TYPE && fileType <= LAST_PLAYLIST_FILE_TYPE);
    }

    public static boolean isDrmFileType(int fileType) {
        return (fileType >= FIRST_DRM_FILE_TYPE && fileType <= LAST_DRM_FILE_TYPE);
    }

    public static MediaFileType getFileType(String path) {
        int lastDot = path.lastIndexOf(".");
        if (lastDot < 0)
            return null;
        return sFileTypeMap.get(path.substring(lastDot + 1).toUpperCase());
    }

    public static boolean isMimeTypeMedia(String mimeType) {
        int fileType = getFileTypeForMimeType(mimeType);
        return isAudioFileType(fileType) || isVideoFileType(fileType)
                || isImageFileType(fileType) || isPlayListFileType(fileType);
    }

    // generates a title based on file name
    public static String getFileTitle(String path) {
        // extract file name after last slash
        int lastSlash = path.lastIndexOf('/');
        if (lastSlash >= 0) {
            lastSlash++;
            if (lastSlash < path.length()) {
                path = path.substring(lastSlash);
            }
        }
        // truncate the file extension (if any)
        int lastDot = path.lastIndexOf('.');
        if (lastDot > 0) {
            path = path.substring(0, lastDot);
        }
        return path;
    }

    public static int getFileTypeForMimeType(String mimeType) {
        Integer value = sMimeTypeMap.get(mimeType);
        return (value == null ? 0 : value);
    }

    public static String getMimeTypeForFile(File file) {
        MediaFileType mediaFileType = getFileType(file.getAbsolutePath());
        return (mediaFileType == null ? null : mediaFileType.mimeType);
    }

    public static String getMimeTypeForFile(String path) {
        MediaFileType mediaFileType = getFileType(path);
        return (mediaFileType == null ? null : mediaFileType.mimeType);
    }

    public static int getFormatCode(String fileName, String mimeType) {
        if (mimeType != null) {
            Integer value = sMimeTypeToFormatMap.get(mimeType);
            if (value != null) {
                return value;
            }
        }
        int lastDot = fileName.lastIndexOf('.');
        if (lastDot > 0) {
            String extension = fileName.substring(lastDot + 1).toUpperCase();
            Integer value = sFileTypeToFormatMap.get(extension);
            if (value != null) {
                return value;
            }
        }
        return MtpConstants.FORMAT_UNDEFINED;
    }

    public static String getMimeTypeForFormatCode(int formatCode) {
        return sFormatToMimeTypeMap.get(formatCode);
    }
}

File: base/src/main/java/com/liabit/util/MediaFile.kt
package com.liabit.util

import android.mtp.MtpConstants
import java.util.*

@Suppress("MemberVisibilityCanBePrivate", "unused")
object MediaFile {

    // maps mime type to MTP format code
    private val mMimeTypeToFormatMap = HashMap<String, Int>()

    // maps MTP format code to mime type
    private val mFormatToMimeTypeMap = HashMap<Int, String>()

    private fun addFileType(mtpFormatCode: Int, mimeType: String) {
        if (!mMimeTypeToFormatMap.containsKey(mimeType)) {
            mMimeTypeToFormatMap[mimeType] = mtpFormatCode
        }
        if (!mFormatToMimeTypeMap.containsKey(mtpFormatCode)) {
            mFormatToMimeTypeMap[mtpFormatCode] = mimeType
        }
    }

    /**
     * The MIME type for data whose type is otherwise unknown.
     *
     *
     * Per RFC 2046, the "application" media type is to be used for discrete
     * data which do not fit in any of the other categories, and the
     * "octet-stream" subtype is used to indicate that a body contains arbitrary
     * binary data.
     */
    private const val MIMETYPE_UNKNOWN = "application/octet-stream"

    /**
     * Format code for HEIF files
     */
    private const val FORMAT_HEIF = 0x3812

    /**
     * Format code for DNG files
     */
    private const val FORMAT_DNG = 0x3811

    /**
     * Format code for unknown image files.
     *
     *
     * Will be used for the formats which are not specified in PTP specification.
     * For instance, WEBP and WBMP.
     */
    private const val FORMAT_DEFINED = 0x3800

    init {
        addFileType(MtpConstants.FORMAT_MP3, "audio/mpeg")
        addFileType(MtpConstants.FORMAT_WAV, "audio/x-wav")
        addFileType(MtpConstants.FORMAT_WMA, "audio/x-ms-wma")
        addFileType(MtpConstants.FORMAT_OGG, "audio/ogg")
        addFileType(MtpConstants.FORMAT_AAC, "audio/aac")
        addFileType(MtpConstants.FORMAT_FLAC, "audio/flac")
        addFileType(MtpConstants.FORMAT_AIFF, "audio/x-aiff")
        addFileType(MtpConstants.FORMAT_MP2, "audio/mpeg")
        addFileType(MtpConstants.FORMAT_MPEG, "video/mpeg")
        addFileType(MtpConstants.FORMAT_MP4_CONTAINER, "video/mp4")
        addFileType(MtpConstants.FORMAT_3GP_CONTAINER, "video/3gpp")
        addFileType(MtpConstants.FORMAT_3GP_CONTAINER, "video/3gpp2")
        addFileType(MtpConstants.FORMAT_AVI, "video/avi")
        addFileType(MtpConstants.FORMAT_WMV, "video/x-ms-wmv")
        addFileType(MtpConstants.FORMAT_ASF, "video/x-ms-asf")
        addFileType(MtpConstants.FORMAT_EXIF_JPEG, "image/jpeg")
        addFileType(MtpConstants.FORMAT_GIF, "image/gif")
        addFileType(MtpConstants.FORMAT_PNG, "image/png")
        addFileType(MtpConstants.FORMAT_BMP, "image/x-ms-bmp")
        addFileType(FORMAT_HEIF, "image/heif")
        addFileType(FORMAT_DNG, "image/x-adobe-dng")
        addFileType(MtpConstants.FORMAT_TIFF, "image/tiff")
        addFileType(MtpConstants.FORMAT_TIFF, "image/x-canon-cr2")
        addFileType(MtpConstants.FORMAT_TIFF, "image/x-nikon-nrw")
        addFileType(MtpConstants.FORMAT_TIFF, "image/x-sony-arw")
        addFileType(MtpConstants.FORMAT_TIFF, "image/x-panasonic-rw2")
        addFileType(MtpConstants.FORMAT_TIFF, "image/x-olympus-orf")
        addFileType(MtpConstants.FORMAT_TIFF, "image/x-pentax-pef")
        addFileType(MtpConstants.FORMAT_TIFF, "image/x-samsung-srw")
        addFileType(MtpConstants.FORMAT_TIFF_EP, "image/tiff")
        addFileType(MtpConstants.FORMAT_TIFF_EP, "image/x-nikon-nef")
        addFileType(MtpConstants.FORMAT_JP2, "image/jp2")
        addFileType(MtpConstants.FORMAT_JPX, "image/jpx")
        addFileType(MtpConstants.FORMAT_M3U_PLAYLIST, "audio/x-mpegurl")
        addFileType(MtpConstants.FORMAT_PLS_PLAYLIST, "audio/x-scpls")
        addFileType(MtpConstants.FORMAT_WPL_PLAYLIST, "application/vnd.ms-wpl")
        addFileType(MtpConstants.FORMAT_ASX_PLAYLIST, "video/x-ms-asf")
        addFileType(MtpConstants.FORMAT_TEXT, "text/plain")
        addFileType(MtpConstants.FORMAT_HTML, "text/html")
        addFileType(MtpConstants.FORMAT_XML_DOCUMENT, "text/xml")
        addFileType(MtpConstants.FORMAT_MS_WORD_DOCUMENT, "application/msword")
        addFileType(MtpConstants.FORMAT_MS_WORD_DOCUMENT, "application/vnd.openxmlformats-officedocument.wordprocessingml.document")
        addFileType(MtpConstants.FORMAT_MS_EXCEL_SPREADSHEET, "application/vnd.ms-excel")
        addFileType(MtpConstants.FORMAT_MS_EXCEL_SPREADSHEET, "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet")
        addFileType(MtpConstants.FORMAT_MS_POWERPOINT_PRESENTATION, "application/vnd.ms-powerpoint")
        addFileType(MtpConstants.FORMAT_MS_POWERPOINT_PRESENTATION, "application/vnd.openxmlformats-officedocument.presentationml.presentation")
    }

    /**
     * Check whether the mime type is document or not.
     *
     * @param mimeType the mime type to check
     * @return true, if the mimeType is matched. Otherwise, false.
     */
    fun isDocumentMimeType(mimeType: String?): Boolean {
        if (mimeType == null) {
            return false
        }
        val normalizedMimeType = normalizeMimeType(mimeType)
        return if (normalizedMimeType.startsWith("text/")) {
            true
        } else when (normalizedMimeType.toLowerCase(Locale.ROOT)) {
            "application/epub+zip",
            "application/msword",
            "application/pdf",
            "application/rtf",
            "application/vnd.ms-excel",
            "application/vnd.ms-excel.addin.macroenabled.12",
            "application/vnd.ms-excel.sheet.binary.macroenabled.12",
            "application/vnd.ms-excel.sheet.macroenabled.12",
            "application/vnd.ms-excel.template.macroenabled.12",
            "application/vnd.ms-powerpoint",
            "application/vnd.ms-powerpoint.addin.macroenabled.12",
            "application/vnd.ms-powerpoint.presentation.macroenabled.12",
            "application/vnd.ms-powerpoint.slideshow.macroenabled.12",
            "application/vnd.ms-powerpoint.template.macroenabled.12",
            "application/vnd.ms-word.document.macroenabled.12",
            "application/vnd.ms-word.template.macroenabled.12",
            "application/vnd.oasis.opendocument.chart",
            "application/vnd.oasis.opendocument.database",
            "application/vnd.oasis.opendocument.formula",
            "application/vnd.oasis.opendocument.graphics",
            "application/vnd.oasis.opendocument.graphics-template",
            "application/vnd.oasis.opendocument.presentation",
            "application/vnd.oasis.opendocument.presentation-template",
            "application/vnd.oasis.opendocument.spreadsheet",
            "application/vnd.oasis.opendocument.spreadsheet-template",
            "application/vnd.oasis.opendocument.text",
            "application/vnd.oasis.opendocument.text-master",
            "application/vnd.oasis.opendocument.text-template",
            "application/vnd.oasis.opendocument.text-web",
            "application/vnd.openxmlformats-officedocument.presentationml.presentation",
            "application/vnd.openxmlformats-officedocument.presentationml.slideshow",
            "application/vnd.openxmlformats-officedocument.presentationml.template",
            "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
            "application/vnd.openxmlformats-officedocument.spreadsheetml.template",
            "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
            "application/vnd.openxmlformats-officedocument.wordprocessingml.template",
            "application/vnd.stardivision.calc",
            "application/vnd.stardivision.chart",
            "application/vnd.stardivision.draw",
            "application/vnd.stardivision.impress",
            "application/vnd.stardivision.impress-packed", "application/vnd.stardivision.mail",
            "application/vnd.stardivision.math",
            "application/vnd.stardivision.writer",
            "application/vnd.stardivision.writer-global", "application/vnd.sun.xml.calc",
            "application/vnd.sun.xml.calc.template",
            "application/vnd.sun.xml.draw",
            "application/vnd.sun.xml.draw.template",
            "application/vnd.sun.xml.impress",
            "application/vnd.sun.xml.impress.template",
            "application/vnd.sun.xml.math",
            "application/vnd.sun.xml.writer",
            "application/vnd.sun.xml.writer.global",
            "application/vnd.sun.xml.writer.template",
            "application/x-mspublisher" -> true
            else -> false
        }
    }

    fun isExifMimeType(mimeType: String?): Boolean {
        // For simplicity, assume that all image files might have EXIF data
        return isImageMimeType(mimeType)
    }

    fun isAudioMimeType(mimeType: String?): Boolean {
        return normalizeMimeType(mimeType).startsWith("audio/")
    }

    fun isVideoMimeType(mimeType: String?): Boolean {
        return normalizeMimeType(mimeType).startsWith("video/")
    }

    fun isImageMimeType(mimeType: String?): Boolean {
        return normalizeMimeType(mimeType).startsWith("image/")
    }

    fun isPlayListMimeType(mimeType: String?): Boolean {
        return when (normalizeMimeType(mimeType)) {
            "application/vnd.ms-wpl",
            "audio/x-mpegurl",
            "audio/mpegurl",
            "application/x-mpegurl",
            "application/vnd.apple.mpegurl",
            "audio/x-scpls" -> true
            else -> false
        }
    }

    fun isDrmMimeType(mimeType: String?): Boolean {
        return normalizeMimeType(mimeType) == "application/x-android-drm-fl"
    }

    // generates a title based on file name
    fun getFileTitle(path: String): String {
        // extract file name after last slash
        var title = path
        var lastSlash = title.lastIndexOf('/')
        if (lastSlash >= 0) {
            lastSlash++
            if (lastSlash < title.length) {
                title = title.substring(lastSlash)
            }
        }
        // truncate the file extension (if any)
        val lastDot = title.lastIndexOf('.')
        if (lastDot > 0) {
            title = title.substring(0, lastDot)
        }
        return title
    }

    fun getFileExtension(path: String?): String? {
        if (path == null) {
            return null
        }
        val lastDot = path.lastIndexOf('.')
        return if (lastDot >= 0) path.substring(lastDot + 1) else null
    }

    /**
     * Find the best MIME type for the given item. Prefers mappings from file
     * extensions, since they're more accurate than format codes.
     */
    fun getMimeType(path: String?, formatCode: Int): String {
        // First look for extension mapping
        val mimeType = getMimeTypeForFile(path)
        // Otherwise look for format mapping
        return if (MIMETYPE_UNKNOWN != mimeType) mimeType else getMimeTypeForFormatCode(formatCode)
    }

    fun getMimeTypeForFile(path: String?): String {
        val ext = getFileExtension(path)
        val mimeType = MimeMap.guessMimeTypeFromExtension(ext)
        return mimeType ?: MIMETYPE_UNKNOWN
    }

    fun getMimeTypeForFormatCode(formatCode: Int): String {
        val mimeType = mFormatToMimeTypeMap[formatCode]
        return mimeType ?: MIMETYPE_UNKNOWN
    }

    /**
     * Find the best MTP format code mapping for the given item. Prefers
     * mappings from MIME types, since they're more accurate than file
     * extensions.
     */
    fun getFormatCode(path: String?, mimeType: String?): Int {
        // First look for MIME type mapping
        val formatCode = getFormatCodeForMimeType(mimeType)
        // Otherwise look for extension mapping
        return if (formatCode != MtpConstants.FORMAT_UNDEFINED) formatCode else getFormatCodeForFile(path)
    }

    fun getFormatCodeForFile(path: String?): Int {
        return getFormatCodeForMimeType(getMimeTypeForFile(path))
    }

    fun getFormatCodeForMimeType(mimeType: String?): Int {
        var mime = mimeType
        if (mime == null) {
            return MtpConstants.FORMAT_UNDEFINED
        }
        // First look for direct mapping
        var value = mMimeTypeToFormatMap[mime]
        if (value != null) {
            return value
        }
        // Otherwise look for indirect mapping
        mime = normalizeMimeType(mime)
        value = mMimeTypeToFormatMap[mime]
        return value ?: when {
            mime.startsWith("audio/") -> MtpConstants.FORMAT_UNDEFINED_AUDIO
            mime.startsWith("video/") -> MtpConstants.FORMAT_UNDEFINED_VIDEO
            mime.startsWith("image/") -> FORMAT_DEFINED
            else -> MtpConstants.FORMAT_UNDEFINED
        }
    }

    /**
     * Normalize the given MIME type by bouncing through a default file
     * extension, if defined. This handles cases like "application/x-flac" to
     * ".flac" to "audio/flac".
     */
    private fun normalizeMimeType(mimeType: String?): String {
        val extension = MimeMap.guessExtensionFromMimeType(mimeType)
        if (extension != null) {
            val extensionMimeType = MimeMap.guessMimeTypeFromExtension(extension)
            if (extensionMimeType != null) {
                return extensionMimeType
            }
        }
        return mimeType ?: MIMETYPE_UNKNOWN
    }

}